python/pyspark/sql/types.py

```diff
@@ -276,7 +276,18 @@ def toInternal(self, dt: datetime.datetime) -> int:
     def fromInternal(self, ts: int) -> datetime.datetime:
         if ts is not None:
             # using int to avoid precision loss in float
-            return datetime.datetime.fromtimestamp(ts // 1000000).replace(microsecond=ts % 1000000)
+            return (
+                datetime.datetime
+                # Set the time zone to UTC because the TIMESTAMP type stores timestamps
+                # as the number of microseconds from the epoch of 1970-01-01T00:00:00.000000Z
+                # in the UTC time zone.
+                .fromtimestamp(ts // 1000000, tz=datetime.timezone.utc)
+                .replace(microsecond=ts % 1000000)
+                # Convert to the local time zone and remove the time zone info to
+                # to have the result type `datetime64[us]` for compatibility.
+                .astimezone(None)
+                .replace(tzinfo=None)
```

Comment:
We should probably fix `toInternal` too. If this shifts from UTC to a local time, `toInternal` should shifts the local time to UTC. Otherwise, roundtrip wouldn't work properly.

Reply:
Let me check this, and add a roundtrip test.

Reply:
shall we remove timezone for timestamp ntz as well?

Reply:
Timestamp ntz should be fine. Scala Java also shift them to UTC if I read the codepath correctly.

Reply:
Here what we need for ltz is set the timezones as utc but don't shitf the offset for the zone

Reply:
yea Java side also shift the timestamp to UTC but its value is timezone-less (`LocalDateTime`).